### PR TITLE
raftstore-v2: implement local read for raftstore-v2

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #12842

What's Changed:

This PR involves:
- Refactor `LocalReader` of raftstore-v1 to make some code sharable with raftstore-v2
- Implement `LocalReader` of raftstore-v2 and add relevant tests
- Replace `callback` with `ReadCallback` trait

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/test_raftstore/src/server.rs [hunk 1]
```diff
@@ -284,7 +284,7 @@ impl ServerCluster {
             }
         }
 
-        let local_reader = LocalReader::new(
+        let local_reader = LocalReaderCore::new(
```
Comment (BusyJay) [inline components/test_raftstore/src/server.rs:287]: Why use `Core`?

File: components/raftstore-v2/src/batch/store.rs [hunk 2]
```diff
@@ -490,3 +496,20 @@ where
     };
     (StoreRouter { router, logger }, system)
 }
+
+pub trait MsgRouter: Send {
```
Comment (BusyJay) [inline components/raftstore-v2/src/batch/store.rs:500]: It seems only used in tests. I suggest to put it to `local.rs`.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/batch/store.rs:500]: Good suggestion.

File: components/raftstore-v2/src/batch/store.rs [hunk 3]
```diff
@@ -467,6 +467,12 @@ impl<EK: KvEngine, ER: RaftEngine> DerefMut for StoreRouter<EK, ER> {
     }
 }
 
+impl<EK: KvEngine, ER: RaftEngine> CasualRouter<EK> for StoreRouter<EK, ER> {
```
Comment (BusyJay) [inline components/raftstore-v2/src/batch/store.rs:470]: Why is this necessary? I think it's the same as send.

File: components/raftstore/src/store/worker/read.rs [hunk 4]
```diff
@@ -673,10 +601,7 @@ where
         rd.filter(|r| !r.pending_remove)
     }
 
-    pub fn pre_propose_raft_command(
-        &mut self,
-        req: &RaftCmdRequest,
-    ) -> Result<Option<(D, RequestPolicy)>> {
+    pub fn request_check(&mut self, req: &RaftCmdRequest) -> Result<Option<D>> {
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:604]: What does `request_check` mean?

Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:604]: Validate request.

File: components/raftstore-v2/src/operation/query/local.rs [hunk 5]
```diff
@@ -29,20 +30,206 @@ use raftstore::{
...
+        }
+
+        // try to renew the lease by sending read query where the reading process may
+        // renew the lease
+        let (msg, mut sub) = PeerMsg::raft_query(req.clone());
+        MsgRouter::send(&self.router, region_id, msg);
+        if let Some(query_res) = sub.result().await {
+            if query_res.read().is_some() {
+                // Query successful, so try again.
+
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:187]: indicate this comment is for this `if branch` rather than the follower line of code.

[the remark this answers is not in the document]
Comment (BusyJay) [reply components/raftstore-v2/src/operation/query/local.rs:187]: Then it should be commented on top of `if`.

File: components/raftstore/src/store/worker/read.rs (hunk 4 above)
Comment (BusyJay) [reply components/raftstore/src/store/worker/read.rs:604]: Bad style, we use `verb + noun` instead of `noun + verb`.

Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs]: Fine.

File: components/raftstore-v2/src/operation/query/local.rs [hunk 6]
```diff
@@ -145,27 +331,107 @@ where
...
+    }
+
+    fn inspect_lease(&mut self) -> LeaseState {
+        // TODO: disable localreader if we did not enable raft's check_quorum.
+        if self.delegate.leader_lease.is_some() {
+            // We skip lease check, because it is postponed until `handle_read`.
+            LeaseState::Valid
+        } else {
+            debug!(self.logger, "rejected by leader lease"; "tag" => &self.delegate.tag);
+            // TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().reject_reason.no_lease.inc());
```
Comment (5kbpers) [inline components/raftstore-v2/src/operation/query/local.rs:401]: Why remove this?

File: components/raftstore-v2/src/operation/query/local.rs [hunk 7]
```diff
@@ -29,20 +30,205 @@ use raftstore::{
...
+
+                    let region = Arc::clone(&delegate.region);
+                    let snap = RegionSnapshot::from_snapshot(
+                        delegate.get_snapshot(None, &mut None),
+                        region,
+                    );
+
+                    delegate.check_stale_read_safe::<E>(read_ts)?;
+
+                    // TLS_LOCAL_READ_METRICS.with(|m|
```
Comment (5kbpers) [inline components/raftstore-v2/src/operation/query/local.rs:150]: (ditto)

File: components/raftstore-v2/src/operation/query/local.rs (hunk 6 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs]: I don't know... It's strange why I remove it.
I have added it back.

File: components/raftstore/src/store/worker/read.rs [hunk 8]
```diff
@@ -538,109 +537,36 @@ impl Progress {
 /// #[RaftstoreCommon]: LocalReader is an entry point where local read requests are dipatch to the
 /// relevant regions by LocalReader so that these requests can be handled by the
 /// relevant ReadDelegate respectively.
-pub struct LocalReader<C, E, D, S>
+pub struct LocalReaderCore<E, D, S>
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:540]: `E` is not needed anymore.

Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs]: Good catch.

File: components/raftstore/src/store/worker/read.rs [hunk 9]
```diff
@@ -437,10 +439,10 @@ impl ReadDelegate {
         false
     }
 
-    pub fn check_stale_read_safe<S: Snapshot>(
+    pub fn check_stale_read_safe<EK: KvEngine>(
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:442]: Wny need a generic?

File: components/raftstore/src/store/worker/read.rs [hunk 10]
```diff
@@ -728,16 +645,116 @@ where
...
+    fn clone(&self) -> Self {
+        LocalReaderCore {
+            store_meta: self.store_meta.clone(),
+            store_id: self.store_id.clone(),
+            delegates: LruCache::with_capacity_and_sample(0, 7),
+        }
+    }
+}
+
+pub struct LocalReader<C, E, D, S>
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:666]: Now that we have two different `LocalReader`, it's meaningless to use generic `D` and `S` anymore.

File: components/raftstore/src/store/worker/read.rs (hunk 9 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs]: It may be used before, but now it is not needed.

File: components/raftstore-v2/src/operation/query/mod.rs [hunk 11]
```diff
@@ -41,7 +41,7 @@ use crate::{
 };
 
 mod lease;
-mod local;
+pub mod local;
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/query/mod.rs:44]: Should not be public.

File: components/raftstore-v2/src/router/imp.rs [hunk 12]
```diff
@@ -1,13 +1,72 @@
...
 }
+
+/// A router that routes messages to the raftstore
+pub struct ServerRaftStoreRouter<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    router: StoreRouter<EK, ER>,
+    local_reader: RefCell<LocalReader<EK, StoreRouter<EK, ER>>>,
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:32]: Why `RefCell`?

File: components/raftstore-v2/src/router/imp.rs [hunk 13]
```diff
@@ -1,13 +1,72 @@
...
+            router,
+            local_reader,
+        }
+    }
+
+    pub fn router(&self) -> &StoreRouter<EK, ER> {
+        &self.router
+    }
+
+    pub fn router_mut(&mut self) -> &mut StoreRouter<EK, ER> {
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:69]: Why need this?

File: components/raftstore-v2/src/router/imp.rs [hunk 14]
```diff
@@ -1,13 +1,72 @@
...
+        logger: Logger,
+    ) -> Self {
+        let local_reader = RefCell::new(LocalReader::new(
+            StoreMetaDelegate::new(store_meta),
+            router.clone(),
+            logger,
+        ));
+        ServerRaftStoreRouter {
+            router,
+            local_reader,
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:61]: It doesn't seem to be used.

File: components/raftstore-v2/src/operation/mod.rs [hunk 15]
```diff
@@ -7,4 +7,5 @@ mod ready;
 
 pub use command::{SimpleWriteDecoder, SimpleWriteEncoder};
 pub use life::DestroyProgress;
+pub use query::local::{CachedReadDelegate, LocalReader, MsgRouter, StoreMetaDelegate};
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/mod.rs:10]: Why export `CachedReadDelegate`, `MsgRouter` and `StoreMetaDelegate`?

File: components/raftstore-v2/src/router/imp.rs (hunk 13 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs:69]: `impl DerefMut for TestRouter` needs this.

File: components/raftstore-v2/src/router/imp.rs (hunk 12 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs:32]: So get_snapshot can be called on & rather than &mut. Is this not needed?

File: components/raftstore-v2/src/router/imp.rs (hunk 14 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs:63]: It is used now.

File: components/raftstore-v2/src/operation/mod.rs (hunk 15 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/mod.rs]: StoreMetaDelegate is needed while the others are not now.

File: components/raftstore-v2/src/fsm/peer.rs [hunk 16]
```diff
@@ -2,19 +2,20 @@
...
 
-use std::borrow::Cow;
+use std::{borrow::Cow, sync::Arc};
 
 use batch_system::{BasicMailbox, Fsm};
 use crossbeam::channel::TryRecvError;
 use engine_traits::{KvEngine, RaftEngine, TabletFactory};
 use kvproto::metapb;
-use raftstore::store::{Config, Transport};
+use raftstore::store::{Config, ReadDelegate, TrackVer, Transport};
```
Comment (BusyJay) [inline components/raftstore-v2/src/fsm/peer.rs:11]: Why import these?

File: components/raftstore-v2/src/operation/query/local.rs [hunk 17]
```diff
@@ -96,7 +271,7 @@ where
 }
 
 #[derive(Clone)]
-struct StoreMetaDelegate<E>
+pub struct StoreMetaDelegate<E>
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/query/local.rs:274]: It's only used in this file, doesn't need to be public.

File: components/raftstore-v2/src/operation/query/mod.rs [hunk 18]
```diff
@@ -41,7 +41,7 @@ use crate::{
 };
 
 mod lease;
-mod local;
+pub(crate) mod local;
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/query/mod.rs:44]: Should re-export internal type instead of making the whole module public.

File: components/raftstore-v2/src/raft/storage.rs [hunk 19]
```diff
@@ -61,6 +64,15 @@ pub struct Storage<ER> {
     logger: Logger,
 }
 
+impl<ER: RaftEngine> Deref for Storage<ER> {
```
Comment (BusyJay) [inline components/raftstore-v2/src/raft/storage.rs:66]: It's not implemented on purpose to avoid confusion.

File: components/raftstore-v2/src/router/imp.rs [hunk 20]
```diff
@@ -1,13 +1,87 @@
...
+    pub fn router_mut(&mut self) -> &mut StoreRouter<EK, ER> {
+        &mut self.router
+    }
+
+    #[allow(clippy::await_holding_refcell_ref)]
+    pub fn get_snapshot(
+        &self,
+        req: RaftCmdRequest,
+    ) -> std::result::Result<RegionSnapshot<EK::Snapshot>, RaftCmdResponse> {
+        block_on({
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:80]: Should be async.

File: components/raftstore-v2/tests/integrations/mod.rs [hunk 21]
```diff
@@ -50,19 +52,19 @@ mod test_read;
...
     fn deref(&self) -> &Self::Target {
-        &self.0
+        self.0.router()
     }
 }
 
 impl DerefMut for TestRouter {
     fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+        self.0.router_mut()
```
Comment (BusyJay) [inline components/raftstore-v2/tests/integrations/mod.rs:67]: Should make target as `ServerRaftStoreRouter` instead of `StoreRouter`. raftstore should only export one type of router.

File: components/raftstore-v2/src/fsm/peer.rs (hunk 16 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/fsm/peer.rs:11]: Residual code, I will remove them.

File: components/raftstore-v2/src/operation/query/local.rs (hunk 17 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:274]: It's also used in router/imp.rs

File: components/raftstore-v2/src/operation/query/mod.rs (hunk 18 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/mod.rs:44]: Godd suggestion.

File: components/raftstore-v2/src/raft/storage.rs (hunk 19 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/raft/storage.rs:66]: Fine.

File: components/raftstore-v2/tests/integrations/mod.rs (hunk 21 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/tests/integrations/mod.rs]: Good suggestion.

File: components/raftstore-v2/src/router/imp.rs (hunk 12 above)
Comment (BusyJay) [reply components/raftstore-v2/src/router/imp.rs:38]: What could be the problem if using `&mut`?

File: components/raftstore-v2/src/raft/storage.rs [hunk 22]
```diff
@@ -1,6 +1,9 @@
 // Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
 
-use std::fmt::{self, Debug, Formatter};
+use std::{
```
Comment (BusyJay) [inline components/raftstore-v2/src/raft/storage.rs:3]: Useless imports.

File: components/raftstore-v2/src/operation/query/local.rs [hunk 23]
```diff
@@ -2,44 +2,234 @@
...
+                // It can not handle other policies.
+                Ok(_) => Ok(None),
+                Err(e) => Err(e),
+            }
+        } else {
+            Err(Error::RegionNotFound(req.get_header().get_region_id()))
+        }
+    }
+
+    fn try_get_snapshot(
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/query/local.rs:101]: Do we only support snapshot here? It seems CmdType::Get is used in some test. Get is just a simple wrapper of snapshot through.

File: components/raftstore-v2/src/operation/query/local.rs [hunk 24]
```diff
@@ -2,44 +2,234 @@
...
+        &mut self,
+        req: RaftCmdRequest,
+    ) -> std::result::Result<Option<RegionSnapshot<E::Snapshot>>, RaftCmdResponse> {
+        match self.pre_propose_raft_command(&req) {
+            Ok(Some((mut delegate, policy))) => match policy {
+                RequestPolicy::ReadLocal => {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/query/local.rs:107]: TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().local_executed_requests.inc()); is missed?

File: components/raftstore/src/store/worker/read.rs [hunk 25]
```diff
@@ -538,109 +533,31 @@ impl Progress {
...
-    D: ReadExecutor<E> + Deref<Target = ReadDelegate>,
-    S: ReadExecutorProvider<E, Executor = D>,
+    D: ReadExecutor + Deref<Target = ReadDelegate>,
+    S: ReadExecutorProvider<Executor = D>,
 {
     pub store_id: Cell<Option<u64>>,
     store_meta: S,
-    kv_engine: E,
     // region id -> ReadDelegate
     // The use of `Arc` here is a workaround, see the comment at `get_delegate`
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/worker/read.rs:544]: // The use of `Arc` here is a workaround, see the comment at `get_delegate`
This comment seems obsolete already. please remove it.

File: components/raftstore-v2/src/router/imp.rs [hunk 26]
```diff
@@ -1,13 +1,81 @@
...
+            router,
+            local_reader,
+        }
+    }
+
+    pub fn send(&self, addr: u64, msg: PeerMsg) -> Result<(), TrySendError<PeerMsg>> {
+        self.router.send(addr, msg)
+    }
+
+    pub fn send_raft_message(
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/router/imp.rs:66]: what's the caller code of this function, same as send()

File: components/raftstore-v2/src/router/imp.rs [hunk 27]
```diff
@@ -1,13 +1,81 @@
...
+
+impl<EK, ER> Clone for ServerRaftStoreRouter<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    fn clone(&self) -> Self {
+        ServerRaftStoreRouter {
+            router: self.router.clone(),
+            local_reader: self.local_reader.clone(),
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/router/imp.rs:44]: if local_reader is being called with borrow_mut(), clone will panic.

File: components/raftstore-v2/src/router/imp.rs [hunk 28]
```diff
@@ -1,13 +1,81 @@
...
+    ) -> std::result::Result<(), TrySendError<Box<RaftMessage>>> {
+        self.router.send_raft_message(msg)
+    }
+
+    #[allow(clippy::await_holding_refcell_ref)]
+    pub async fn get_snapshot(
+        &self,
+        req: RaftCmdRequest,
+    ) -> std::result::Result<RegionSnapshot<EK::Snapshot>, RaftCmdResponse> {
+        let mut local_reader = self.local_reader.borrow_mut();
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/router/imp.rs:78]: self.local_reader.borrow_mut() will panic if it's called concurrently.

File: components/raftstore-v2/src/router/imp.rs (hunk 12 above)
Comment (tonyxuqqi) [reply components/raftstore-v2/src/router/imp.rs:33]: > So get_snapshot can be called on & rather than &mut. Is this not needed?

I think either make get_snapshot use &mut or make the underlying get_snapshot accept &. To me both way is fine, the question is which one needs less code change. Using RefCell here is quite dangerous  as it's easily mis-used and cause panic.

File: components/raftstore-v2/src/operation/query/local.rs [hunk 29]
```diff
@@ -2,44 +2,234 @@
...
+        }
+
+        let mut resp = RaftCmdResponse::default();
+        resp.mut_header().set_error(err);
+        Err(resp)
+    }
+
+    // If the remote lease will be expired in near future send message
+    // to `raftstore` to renew it
+    fn maybe_renew_lease_in_advance(
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:216]: You mean even after the filter of `need_renew_lease` ?

File: components/raftstore-v2/src/operation/query/local.rs (hunk 24 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:107]: Good catch.

File: components/raftstore/src/store/worker/read.rs (hunk 25 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:544]: Good catch.

File: components/raftstore-v2/src/router/imp.rs (hunk 26 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs:66]: It's already used in some tests.

File: components/raftstore-v2/src/router/imp.rs (hunk 12 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs:33]: So, why v1 uses `RefCell`?

File: components/raftstore-v2/src/operation/query/local.rs [hunk 30]
```diff
@@ -2,44 +2,234 @@
...
+    ) {
+        if !delegate.need_renew_lease(ts) {
+            return;
+        }
+
+        let region_id = req.header.get_ref().region_id;
+        TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().renew_lease_advance.inc());
+        // Send a read query which may renew the lease
+        let (msg, sub) = PeerMsg::raft_query(req.clone());
+        let _ = MsgRouter::send(&self.router, region_id, msg);
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:230]: Okay.

File: components/raftstore-v2/src/router/imp.rs (hunk 28 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs:78]: It may be the same issue in v1 as well?

File: components/raftstore-v2/src/operation/query/local.rs (hunk 23 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:101]: Yes, we may remove CmdType::Get later in some time.

File: components/raftstore-v2/src/router/imp.rs (hunk 12 above)
Comment (BusyJay) [reply components/raftstore-v2/src/router/imp.rs:33]: You may also try to get rid of `RefCell` for v1.

File: components/raftstore-v2/src/router/imp.rs [hunk 31]
```diff
@@ -1,13 +1,78 @@
 // Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
 
+use std::{
+    cell::RefCell,
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:4]: Useless import.

File: components/raftstore-v2/src/router/imp.rs [hunk 32]
```diff
@@ -1,13 +1,78 @@
...
+use crate::{batch::StoreRouter, operation::LocalReader, StoreMeta};
 
 impl<EK: KvEngine, ER: RaftEngine> LogFetchedNotifier for StoreRouter<EK, ER> {
     fn notify(&self, region_id: u64, fetched: FetchedLogs) {
         let _ = self.force_send(region_id, PeerMsg::FetchedLogs(fetched));
     }
 }
+
+/// A router that routes messages to the raftstore
+pub struct ServerRaftStoreRouter<EK, ER>
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:27]:
```suggestion
pub struct RaftRouter<EK, ER>
```

File: components/raftstore-v2/src/router/imp.rs [hunk 33]
```diff
@@ -1,13 +1,78 @@
...
+            local_reader: self.local_reader.clone(),
+        }
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> ServerRaftStoreRouter<EK, ER> {
+    pub fn new(
+        store_meta: Arc<Mutex<StoreMeta<EK>>>,
+        router: StoreRouter<EK, ER>,
+        logger: Logger,
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:53]: logger should be taken from `StoreRouter` instead.

File: components/raftstore-v2/src/router/imp.rs [hunk 34]
```diff
@@ -1,13 +1,78 @@
...
+        ServerRaftStoreRouter {
+            router: self.router.clone(),
+            local_reader: self.local_reader.clone(),
+        }
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> ServerRaftStoreRouter<EK, ER> {
+    pub fn new(
+        store_meta: Arc<Mutex<StoreMeta<EK>>>,
```
Comment (BusyJay) [inline components/raftstore-v2/src/router/imp.rs:51]: It's strange to pass the parameter from outside. `store_meta` is expected to be created and maintained by raftstore.

File: components/raftstore-v2/src/raft/peer.rs [hunk 35]
```diff
@@ -382,4 +374,25 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
     pub fn proposals_mut(&mut self) -> &mut ProposalQueue<Vec<CmdResChannel>> {
         &mut self.proposals
     }
+
+    pub fn generate_read_delegate(&self) -> ReadDelegate {
+        let region = self.region().clone();
+        let region_id = region.get_id();
+        let peer_id = self.peer().get_id();
+        ReadDelegate {
```
Comment (BusyJay) [inline components/raftstore-v2/src/raft/peer.rs:382]: Better create a constructor for `ReadDelegate`.

File: components/raftstore-v2/src/operation/query/local.rs [hunk 36]
```diff
@@ -142,39 +345,339 @@ where
     }
 }
 
+struct SnapRequestInspector<'r> {
+    delegate: &'r ReadDelegate,
+    logger: &'r Logger,
+}
+
+impl<'r> RequestInspector for SnapRequestInspector<'r> {
+    fn inspect(&mut self, req: &RaftCmdRequest) -> Result<RequestPolicy> {
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/query/local.rs:354]: What's the point of implementing `RequestInspector`?

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:354]: Remove unnecessary checks, and assert for the snap request.

File: components/raftstore-v2/src/router/imp.rs (hunk 31 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs:4]: Good catch.

File: components/raftstore-v2/src/operation/query/local.rs (hunk 36 above)
Comment (BusyJay) [reply components/raftstore-v2/src/operation/query/local.rs:354]: What's the difference between implementing as methods instead of trait?

File: components/raftstore-v2/src/raft/peer.rs (hunk 35 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/raft/peer.rs]: Done.

File: components/raftstore-v2/src/router/imp.rs (hunk 33 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/router/imp.rs]: Done.

File: components/raftstore-v2/src/operation/mod.rs [hunk 37]
```diff
@@ -8,3 +8,5 @@ mod ready;
 pub use command::{SimpleWriteDecoder, SimpleWriteEncoder};
 pub use life::DestroyProgress;
 pub use ready::AsyncWriter;
+
+pub(crate) use self::query::LocalReader;
```
Comment (5kbpers) [inline components/raftstore-v2/src/operation/mod.rs:12]: If it could be used by other mods, using `pub` would be better.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/mod.rs:12]: It's supposed not to be used by other module.

**Review by 5kbpers — APPROVED** [review]

File: components/raftstore-v2/src/operation/query/local.rs (hunk 29 above)
[the remark this answers is not in the document]
Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/query/local.rs:226]: > You mean even after the filter of `need_renew_lease` ?
Yes. So There're two scenarios:
1) The region is active and 1K read requests hit read local and the lease is about expire. need_renew_lease will return true for all these requests. And they're all sent to Raftstore to renew the lease. 
2) The region is inactive and 1K read requests hit read index and they're all sent to Raftstore to renew the lease. 
The ideal result is that for 1) only one pre-renew lease is sent per region. And for 2), there's also only one renew lease needed, other requests just need to wait for its finish.

File: components/raftstore-v2/src/operation/query/local.rs [hunk 38]
```diff
@@ -2,44 +2,251 @@
...
+    // renew the lease
+    async fn try_to_renew_lease(
+        &self,
+        region_id: u64,
+        req: &RaftCmdRequest,
+    ) -> std::result::Result<Option<QueryResult>, RaftCmdResponse> {
+        let (msg, sub) = PeerMsg::raft_query(req.clone());
+        let mut err = errorpb::Error::default();
+        match MsgRouter::send(&self.router, region_id, msg) {
+            Ok(()) => return Ok(sub.result().await),
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/query/local.rs:205]: I'm wondering if we can clone and cache sub here. 
So when sub exists,  we don't need to issue the MsgRouter::send again for later requests that are before sub.result().await return . Just wait on sub.result().await. And once await returns, clear the cache. 
This is to dedup unnecessary renew lease requests to raftstore. 
With this, it can even remove the read index amend logic in raftstore v2. Because redundant read index messages will be greatly reduced or in fact completely removed. @BusyJay What do you think?

File: components/raftstore-v2/src/operation/query/local.rs [hunk 39]
```diff
@@ -2,44 +2,251 @@
...
+        ts: Timespec,
+    ) {
+        if !delegate.need_renew_lease(ts) {
+            return;
+        }
+
+        let region_id = req.header.get_ref().region_id;
+        TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().renew_lease_advance.inc());
+        // Send a read query which may renew the lease
+        let (msg, sub) = PeerMsg::raft_query(req.clone());
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/query/local.rs:239]: Same as above. If we cache the sub after send successfully, it can reduce the unnecessary renew lease requests.

File: components/raftstore-v2/src/operation/mod.rs (hunk 37 above)
Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/mod.rs:12]: In v1's code, it's called in components/server as well.

File: components/raftstore-v2/src/operation/query/local.rs (hunk 38 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/query/local.rs:205]: If it is possible, I think we can propose another PR to implement it and see the improvements.

**Review by tonyxuqqi — APPROVED** [review]

**Comment by BusyJay** [comment]
Why merge this? I think it depends on #13568

File: components/raftstore-v2/src/operation/query/local.rs (hunk 38 above)
Comment (BusyJay) [reply components/raftstore-v2/src/operation/query/local.rs:205]: ReadIndex amend works for requests across threads.

**Comment by SpadeA-Tang** [comment]
> Why merge this? I think it depends on #13568

The circumstance here is much easier. It only has snap oepration and does not have cache.

**Comment by BusyJay** [comment]
> The circumstance here is much easier. It only has snap oepration and does not have cache.

It doesn't have cross snapshot cache, but it may has cache within the same region. And this PR also lacks the check of last_valid_ts.

**Comment by SpadeA-Tang** [comment]
> 

Checking last_valid_ts means we have to acquire the system time twice? One before getting the snapshot and one after it.

**Comment by SpadeA-Tang** [comment]
> Checking last_valid_ts means we have to acquire the system time twice? One before getting the snapshot and one after it.

Checking last_valid_ts means we have to acquire the system time twice? One before getting the snapshot and one after it.

**Comment by SpadeA-Tang** [comment]
> > The circumstance here is much easier. It only has snap oepration and does not have cache.
> 
> It doesn't have cross snapshot cache, but it may has cache within the same region. And this PR also lacks the check of last_valid_ts.

Checking last_valid_ts means we have to acquire the system time twice? One before getting the snapshot and one after it.

**Comment by BusyJay** [comment]
> Checking last_valid_ts means we have to acquire the system time twice? One before getting the snapshot and one after it.

Yes, for V1. For v2, only check if latest tablet is `None` is sufficient.
